src/resolved.ts

```diff
@@ -45,13 +46,32 @@ export class Resolved {
 
     let piece = this.unresolved;
 
-    for (let i = 0; i < path.length; i++) {
+    for (const segment of path) {
       if (!isObject(piece)) return false;
 
-      if (path[i] in piece) {
-        piece = piece[path[i]];
+      if (segment in piece) {
+        piece = piece[segment];
       } else if (hasRef(piece)) {
-        return this.doesBelongToDoc([...pointerToPath(piece.$ref), ...path.slice(i)]);
+        if (this.spec.source === void 0) return false;
+        let nodeData;
+        try {
+          nodeData = this.graph.getNodeData(this.spec.source);
+        } catch {
+          return false;
+        }
+
+        const { refMap } = nodeData;
+        let { $ref } = piece;
+
+        while ($ref in refMap) {
+          $ref = refMap[$ref];
+        }
```

Comment:
This is an odd looking piece of code, could you add a comment describing what this is for?

Reply:
Or just in general maybe this entire `hasRef` logic tree (line 55-74) could be pulled out to its own function, with a comment describing what it does (or some descriptive tests)?